### PR TITLE
Fix zoom and slide on native for PDFView

### PR DESCRIPTION
**Pull request by alexxxwork**
### Details

The carousel component doesn't swipe with single tap when displaying a PDF. User should be able to switch attachments with swipe and to zoom and pan pdf attachments. 
This PR adds a callback to get scale from PDFView component to know if we are in a zoomed state.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15988
PROPOSAL: https://github.com/Expensify/App/issues/15988#issuecomment-1482739389

### Tests

1. Go to any chat
2. Send Image, then send a pdf file
3. Preview image, slide right using fingers (not the right arrow slider) 
4. Verify the pdf displays.
5. Slide left from pdf and slide back right. 
6. Verify that attachments change smoothly 
7. Pinch-zoom or double-tap to zoom pdf
8. Verify that you can move the document with a finger touch and drag
9. Verify that you're not able to switch to other attachment in zoomed state

- [ ] Verify that no errors appear in the JS console

### QA Steps

1. Go to any chat
2. Send Image, then send a pdf file
3. Preview image, slide right using fingers (not the right arrow slider) 
4. Verify the pdf displays.
5. Slide left from pdf and slide back right. 
6. Verify that attachments change smoothly 
7. Pinch-zoom or double-tap to zoom pdf
8. Verify that you can move the document with a finger touch and drag
9. Verify that you're not able to switch to other attachment in zoomed state

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3872759/233027410-88503cd7-5137-4cbc-8215-ad02c1eb1ced.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3872759/233029999-8ea9a55b-af50-400b-84d5-9450b96c6038.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3872759/233027300-f84fa818-ae81-4f62-9b9f-4234d60f8b9d.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3872759/233027522-8366b282-291a-4326-94ae-b4fd236cf67c.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3872759/233019618-055c8c8f-7788-474a-9575-92c15b50d816.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3872759/233019444-0bf021e8-2512-4cf5-a315-a1cb036853a9.mp4
</details>

### CODE REVIEW

File: src/components/AttachmentCarousel/index.js [hunk 1]
```diff
@@ -292,7 +303,8 @@ class AttachmentCarousel extends React.Component {
 
                         // Enable scrolling by swiping on mobile (touch) devices only
                         // disable scroll for desktop/browsers because they add their scrollbars
-                        scrollEnabled={this.canUseTouchScreen}
+                        // disable scroll in PDF zoomed state
```
Comment (luacmartins) [inline src/components/AttachmentCarousel/index.js:306]: the comment says that we disable scroll on zoomed pdfs, but the code seems to disable it when it's not zoomed. Can we make this consistent?

**Review by luacmartins — COMMENTED** [review]
Left a small comment.

File: src/components/AttachmentCarousel/index.js (hunk 1 above)
Comment (alexxxwork) [reply src/components/AttachmentCarousel/index.js:306]: The code enables scroll (scrolling the FlatList with different attachments horizontally) when we are not in a zoomed state. And disables scroll (changing attachments) in a zoomed state. There could be a misunderstanding on what is to scroll in this context.

**Review by 0xmiroslav — COMMENTED** [review]
It's a good idea to move onPress event from `renderCell` to `renderItem` but unfortunately it causes regression of prev/next arrow buttons visibility not changing when tap in these cases:
- while loading pdf
- on password protected pdf
- on image

File: src/components/AttachmentCarousel/index.js [hunk 2]
```diff
@@ -184,10 +219,11 @@ class AttachmentCarousel extends React.Component {
             return;
         }
 
+        const isZoomed = false;
         const page = entry.index;
         const {source, file} = this.getAttachment(entry.item);
         this.props.onNavigate({source: addEncryptedAuthTokenToURL(source), file});
-        this.setState({page, source});
+        this.setState({page, source, isZoomed});
```
[the remark this answers is not in the document]
Comment (alexxxwork) [reply src/components/AttachmentCarousel/index.js:231]: Feel like we should better leave it. There could be issues with arrow navigation from zoomed state on an iPad with keyboard. See [this](https://github.com/Expensify/App/pull/14767) PR

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: src/components/AttachmentView.js [hunk 3]
```diff
@@ -70,49 +81,57 @@ const AttachmentView = (props) => {
...
         return (
-            <ImageView onPress={props.onPress} url={props.source} isAuthTokenRequired={props.isAuthTokenRequired} />
+            <Pressable onPress={props.onPress} style={containerStyles}>
+                <ImageView url={props.source} isAuthTokenRequired={props.isAuthTokenRequired} />
+            </Pressable>
         );
     }
 
     return (
         <View
```
[the remark this answers is not in the document]
Comment (alexxxwork) [reply src/components/AttachmentView.js:106]: I don't think we should remove Pressable in ImageView. I'll just get onPress work in ImageView

**Comment by luacmartins** [comment]
@alexxxwork just one last change requested - https://github.com/Expensify/App/pull/17647#discussion_r1174725308

**Comment by alexxxwork** [comment]
> @alexxxwork just one last change requested - [#17647 (comment)](https://github.com/Expensify/App/pull/17647#discussion_r1174725308)
@luacmartins Changes pushed.

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]
@alexxxwork Please apply my suggested changes. We don't need hand cursor at all on non-touchable devices.

**Review by 0xmiroslav — APPROVED** [review]
Latest version looks good.
Tested all possible cases.
There's dedicated issue https://github.com/Expensify/App/issues/17720 for image zoomed case.
There are some known bugs below but they already happen on `main` so out of scope. I will report them in slack soon.
- weird behavior on clicking left/right arrow on mChrome, similar to https://github.com/Expensify/App/issues/17723#issuecomment-1519838172
- resizing window on web hides Send button on attachment preview
- scroll bar not fully covers image when zoomed
- app freezes when click android device back button on mChrome

All yours @luacmartins

**Review by luacmartins — APPROVED** [review]

File: src/components/AttachmentView.js [hunk 4]
```diff
@@ -69,21 +76,35 @@ const AttachmentView = (props) => {
...
                 sourceURL={sourceURL}
                 style={styles.imageModalPDF}
                 onToggleKeyboard={props.onToggleKeyboard}
+                onScaleChanged={props.onScaleChanged}
+                onLoadComplete={() => !loadComplete && setLoadComplete(true)}
             />
         );
+        return (
+            props.onPress ? (
+                <Pressable onPress={props.onPress} disabled={loadComplete} style={containerStyles}>
```
[the remark this answers is not in the document]
Comment (alexxxwork) [reply src/components/AttachmentView.js:91]: Sorry, completely missed this mention. If I remember this right, there was a specific case when arrows wouldn't hide/show if you press attachment before PDF loaded. So we use Pressable and disable it on PDF load complete.
cc: @0xmiroslav

**Pull request merged**